nat: Do not inc error metric from CT GC if related NAT entry not found

soumynathan opened the pull request:
This PR adds in an additional param to bpf.Map Delete() so that the
error metric count increment can be controlled if the entry not found.

Fixes: #11485
Signed-off-by: Swaminathan Vasudevan <svasudevan@suse.com>

aanm reviewed (changes requested):
It seems to me that `ignoreMetricCountifNotAvailable` is always false on all cases?

brb commented on pkg/bpf/map_linux.go:
```diff
@@ -863,7 +863,7 @@ func (m *Map) deleteCacheEntry(key MapKey, err error) {
 
 // Delete deletes the map entry corresponding to the given key.
-func (m *Map) Delete(key MapKey) error {
+func (m *Map) Delete(key MapKey, ignoreMetricCountifNotAvailable bool) error {
```
Instead of changing the `Delete` method, have you considered creating a new method `DeleteWithIgnoreFoo` which would be then used by the CT? This approach would not require changing all the existing invocations of `Delete`?

soumynathan replied:
@brb good point, let me consider this. My initial understanding is to add an argument to the Delete. But I did realize that I am touching too many files to make this change and it is worth while to just add a new function.

Commit:
nat: Do not inc error metric from CT GC if related NAT entry not found

This PR adds in an additional 'DeleteIgnoreMetricErrorCount' to bpf.Map
so that the error metric count increment can be controlled if entry not
found.

Fixes: #11485
Signed-off-by: Swaminathan Vasudevan <svasudevan@suse.com>

brb commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -294,7 +294,7 @@ func newMap(mapName string, m mapType) *Map {
 
 func purgeCtEntry6(m *Map, key CtKey, natMap NatMap) error {
-	err := m.Delete(key)
+	err := m.DeleteIgnoreMetricErrorCount(key, true)
```
You probably want to use the method at the line 299 instead.

soumynathan replied:
@brb I think I am confused here. I thought you want me to replace the 'DeleteIgnoreMetricErrorCount' function with 'DeleteMapping' which is at line 299.
Can you clarify here.

brb commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -374,7 +374,7 @@ func doGC6(m *Map, filter *GCFilter) gcStats {
 
 func purgeCtEntry4(m *Map, key CtKey, natMap NatMap) error {
-	err := m.Delete(key)
+	err := m.DeleteIgnoreMetricErrorCount(key, true)
```
Same here (line 379).

brb commented on pkg/bpf/map_linux.go:
```diff
@@ -884,6 +884,31 @@ func (m *Map) Delete(key MapKey) error {
 	return err
 }
 
+// DeleteIgnoreMetricErrorCount deletes the map entry without incrementing the
```
I'd include an explanation for when it can happen (when element doesn't exist).

brb commented on pkg/bpf/map_linux.go:
```diff
@@ -884,6 +884,31 @@ func (m *Map) Delete(key MapKey) error {
 
+// DeleteIgnoreMetricErrorCount deletes the map entry without incrementing the
+// metric error count for the given key.
+func (m *Map) DeleteIgnoreMetricErrorCount(key MapKey, ignoreMetricErrorCount bool) error {
```
IMHO, this method name reads as "this method deletes an ignore metric error counter".

joestringer replied:
Should we call it something like `SilentDelete()` or something because it silences metrics which could be used to figure out what it's doing?

brb commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -294,10 +294,7 @@ func newMap(mapName string, m mapType) *Map {
-	if err == nil && natMap != nil {
-		natMap.DeleteMapping(key.GetTupleKey())
-	}
+	err := natMap.DeleteMapping(key.GetTupleKey())
```
Why did you remove the CT entry removal?

soumynathan replied:
Sorry the comment was not clear. I should have asked for clarity.

qmonnet commented on pkg/bpf/map_linux.go:
```diff
@@ -890,6 +890,31 @@ func (m *Map) Delete(key MapKey) error {
+
+	_, errno := deleteElement(m.fd, key.GetKeyPtr())
+	if option.Config.MetricsConfig.BPFMapOps {
+		if !ignoreMetricErrorCount {
+			metrics.BPFMapOps.WithLabelValues(m.commonName(), metricOpDelete, metrics.Errno2Outcome(errno)).Inc()
+		}
```
This `if` condition seems to be the only difference with the original `Delete()` method. It makes little sense to have this chunk at all, since each time you call `DeleteButIgnoreMetricErrorCountInc()` you pass `true` as second argument.

As I understand the initial idea was to add an argument to `Delete()` itself, but that would involve changing the numerous calls to the method, so it's simpler to add a new function. But you can probably base `Delete()` on the new function instead, to avoid having too much duplicated code:

``` go
func (m *Map) DeleteButIgnoreMetricErrorCountInc(key MapKey, ignoreMetricErrorCount bool) error {

	[...]

	if option.Config.MetricsConfig.BPFMapOps && !ignoreMetricErrorCount {
		metrics.BPFMapOps.WithLabelValues(m.commonName(), metricOpDelete, metrics.Errno2Outcome(errno)).Inc()
	}

	[...]

}

func (m *Map) Delete(key MapKey) error {
	return DeleteButIgnoreMetricErrorCountInc(key, false)
}
```

In that case the boolean argument makes sense, and we avoid replicating the code.

Another note: as I understand, in the case we want to skip the metric update, we only want to skip it if the error is `ENOENT`, to avoid silencing errors other than the key not being found. So you might be missing a check here.

Please make sure to address the other suggestions from older comments (e.g. on function name).

joestringer replied:
@qmonnet I think this makes sense. FYI this looks like a "request changes" type feedback, but I see the "eye" icon from a review that selected to "Submit general feedback without explicit approval." Typically we try to avoid using this option to make it clear when changes should be addressed prior to merging, otherwise contributors do not get a clear signal whether action is required or not, and janitors do not get a clear signal about whether it is OK to merge. I've replied below requesting changes partially based on this and partially on the other feedback above.

joestringer replied:
(If this is just Github UI mismatch or my misintepretation, feel free to disregard my comment but just trying to bring some clarification to the state of this PR and suggestions to improve the review process for both contributors and maintainers :-) )

qmonnet replied:
@joestringer No problem and agreed, I can't remember why I went for “comment” on this particular review. Likely I went too fast :/. Thanks for the notice :)

joestringer reviewed (changes requested):
I did not review but scanning through the page I see several unresolved comments.

Requesting changes, please re-request review when they have been addressed.
